### PR TITLE
Check if SSL is used

### PR DESCRIPTION
**Pull request by schwarzkrieger**
Check if connection is secured by using SSL and issue warning message in the Dashboard.

### CODE REVIEW

File: tcms/core/tests/test_views.py [hunk 1]
```diff
@@ -28,15 +28,26 @@ def setUpTestData(cls):
...
-            "doc_url": doc_url,
+            "doc_url": doc_url_base,
             "admin_url": reverse("admin:sites_site_change", args=[settings.SITE_ID]),
         }
+        doc_url_ssl = (
+            "https://kiwitcms.readthedocs.io/en/latest/installing_docker.html"
+            "#ssl-configuration"
+        )
+        cls.ssl_error_message = _(
+            "You are not using secure connection (SSL). "
```
Comment (atodorov) [inline tcms/core/tests/test_views.py:48]: Change to "You are not using a secure connection. See documentation and enable SSL."

File: tcms/core/tests/test_views.py [hunk 2]
```diff
@@ -82,6 +93,22 @@ def test_check_base_url_configured(self):
             response = self.client.get("/", follow=True)
             self.assertNotContains(response, self.base_url_error_message)
 
+    @unittest.skipUnless(
+        os.getenv("TEST_DASHBOARD_CHECK_SSL"),
+        "Checking SSL is not enabled",
+    )
+    def test_check_connection_uses_ssl(self):
+        response = self.client.get("/", follow=True)
+        self.assertNotContains(response, self.ssl_error_message)
```
Comment (atodorov) [inline tcms/core/tests/test_views.py:102]: Remove this scenario. Test server does not support SSL.

File: tcms/core/tests/test_views.py [hunk 3]
```diff
@@ -82,6 +93,22 @@ def test_check_base_url_configured(self):
...
+
+    @unittest.skipUnless(
+        not os.getenv("TEST_DASHBOARD_CHECK_SSL"),
+        "Check not using SSL is not enabled",
+    )
```
Comment (atodorov) [inline tcms/core/tests/test_views.py:107]: Remove the conditional decorator. This test case should be executed unconditionally.

File: tcms/core/tests/test_views.py [hunk 4]
```diff
@@ -28,15 +28,26 @@ def setUpTestData(cls):
...
+            "https://kiwitcms.readthedocs.io/en/latest/admin.html"
+            "#configure-kiwi-s-base-url"
+        )
         cls.base_url_error_message = _(
             "Base URL is not configured! "
             'See <a href="%(doc_url)s">documentation</a> and '
             '<a href="%(admin_url)s">change it</a>'
         ) % {
-            "doc_url": doc_url,
+            "doc_url": doc_url_base,
```
Comment (atodorov) [inline tcms/core/tests/test_views.py:40]: Why do we need variable rename?

File: tcms/core/tests/test_views.py [hunk 5]
```diff
@@ -82,6 +93,22 @@ def test_check_base_url_configured(self):
...
+        response = self.client.get("/", follow=True)
+        self.assertNotContains(response, self.ssl_error_message)
+
+    @unittest.skipUnless(
+        not os.getenv("TEST_DASHBOARD_CHECK_SSL"),
+        "Check not using SSL is not enabled",
+    )
+    def test_check_connection_not_using_ssl(self):
+        response = self.client.get("/", follow=True)
+        self.assertContains(response, self.ssl_error_message)
```
Comment (atodorov) [inline tcms/core/tests/test_views.py:110]: Move the expected message here. It's not used anywhere else.

File: tcms/core/views.py [hunk 6]
```diff
@@ -78,6 +78,26 @@ def get_context_data(self, **kwargs):
...
 
+        # Check for SSL usage
+        doc_url = (
+            "https://kiwitcms.readthedocs.io/en/latest/"
+            "installing_docker.html#ssl-configuration"
+        )
+        if not self.request.is_secure():
+            messages.add_message(
+                self.request,
+                messages.ERROR,
```
Comment (atodorov) [inline tcms/core/views.py:89]: Change to messages.WARNING.

File: tcms/core/views.py [hunk 7]
```diff
@@ -78,6 +78,26 @@ def get_context_data(self, **kwargs):
...
+                    _(
+                        "You are not using secure connection (SSL). "
+                        'See <a href="%(doc_url)s">documentation</a>'
+                    )
+                    % {
+                        "doc_url": doc_url,
+                    }
+                ),
+            )
+
```
Comment (atodorov) [inline tcms/core/views.py:100]: See `tests/test_http.sh` and add additional checks for the error message with http/https.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/core/tests/test_views.py [hunk 8]
```diff
@@ -28,15 +28,26 @@ def setUpTestData(cls):
...
             'See <a href="%(doc_url)s">documentation</a> and '
             '<a href="%(admin_url)s">change it</a>'
         ) % {
-            "doc_url": doc_url,
+            "doc_url": doc_url_base,
             "admin_url": reverse("admin:sites_site_change", args=[settings.SITE_ID]),
         }
+        doc_url_ssl = (
+            "https://kiwitcms.readthedocs.io/en/latest/installing_docker.html"
+            "#ssl-configuration"
```
Comment (atodorov) [inline tcms/core/tests/test_views.py:45]: This section in the docs will also need a warning stating that we're going to issue a warning message in v10.4 onwards.

File: docs/source/installing_docker.rst [hunk 9]
```diff
@@ -184,6 +185,12 @@ bind-mount your own SSL directory to ``/Kiwi/ssl`` inside the docker container!
 More information about generating your own self-signed certificates can be
 found at https://github.com/sgallagher/sscg#full-usage.
 
+.. warning::
+
+    The system will issue a warning if the connection is not secured by SSL!!!
```
Comment (atodorov) [inline docs/source/installing_docker.rst:190]: "The system" -> "Kiwi TCMS"

File: tcms/core/tests/test_views.py (hunk 4 above)
Comment (atodorov) [reply tcms/core/tests/test_views.py:40]: ping ^^^

File: tcms/core/views.py (hunk 7 above)
Comment (atodorov) [reply tcms/core/views.py:100]: ping ^^^

**Review by atodorov — CHANGES_REQUESTED** [review]
See code review comments.

File: docs/source/installing_docker.rst [hunk 10]
```diff
@@ -184,6 +185,12 @@ bind-mount your own SSL directory to ``/Kiwi/ssl`` inside the docker container!
 More information about generating your own self-signed certificates can be
 found at https://github.com/sgallagher/sscg#full-usage.
 
+.. warning::
+
+    KiwiTCMS will issue a warning if the connection is not secured by SSL!!!
```
Comment (atodorov) [inline docs/source/installing_docker.rst:190]: "KiwiTCMS" -> "Kiwi TCMS", don't change project name spelling.

**Review by atodorov — CHANGES_REQUESTED** [review]
Minor change requested. Still needs the tests for http/https. As discussed using robot framework should make this easier.

File: tests/test_http.sh [hunk 11]
```diff
@@ -24,6 +24,20 @@ rlJournalStart
         assert_up_and_running
     rlPhaseEnd
 
+    rlPhaseStartTest "Check for SSL warning message"
+        rlRun -t -c "docker exec -i kiwi_web /Kiwi/manage.py createsuperuser \
+            --username testadmin --email testadmin@domain.com --noinput"
+        rlRun -t -c "docker exec -i kiwi_web /Kiwi/manage.py set_fake_passwords"
```
Comment (atodorov) [inline tests/test_http.sh:30]: `STDERR: CommandError: Only available in debug mode`, see
https://github.com/kiwitcms/enterprise/blob/master/.github/workflows/testing.yml#L146

File: tests/set_testadmin_pass.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+"""
+    Allow testadmin to log in by setting its password
+"""
+from django.contrib.auth.models import User
+
+testadmin = User.objects.get(username='testadmin')
+testadmin.set_password('password')
+testadmin.save()
```
Comment (atodorov) [inline tests/set_testadmin_pass.py:8]: Missing newline at EOF.

File: tests/test_http.sh [hunk 13]
```diff
@@ -27,19 +27,23 @@ rlJournalStart
     rlPhaseStartTest "Check for SSL warning message"
```
Comment (atodorov) [inline tests/test_http.sh:27]: Naming -> Should not display SSL warning for HTTPS connection

File: tests/test_http.sh [hunk 14]
```diff
@@ -27,19 +27,23 @@ rlJournalStart
...
         rm -f /tmp/testcookies.txt
+        rm -f /tmp/testdata.sslon
         rlRun -t -c "curl -k -L -o- -c /tmp/testcookies.txt https://$IP_ADDRESS:8443/"
         CSRF_TOKEN=`grep csrftoken /tmp/testcookies.txt | cut -f 7`
         rlRun -t -c "curl -e https://$IP_ADDRESS:8443/ -d username=testadmin \
             -d password=password -d csrfmiddlewaretoken=$CSRF_TOKEN -k -L -o /tmp/testdata.sslon \
             -b /tmp/testcookies.txt https://$IP_ADDRESS:8443/accounts/login/"
+        rlAssertGrep "<title>Kiwi TCMS - Dashboard</title>" /tmp/testdata.sslon
         rlAssertNotGrep "You are not using a secure connection." /tmp/testdata.sslon
     rlPhaseEnd
```
Comment (atodorov) [inline tests/test_http.sh:41]: You can extract the gist of this section and add another rlPhase to check that the warning message exists when using HTTP only. See line 21 for hints.

**Review by atodorov — CHANGES_REQUESTED** [review]
Minor details need changing. You can use the existing code base and add similar scenario for HTTP.

File: tests/test_http.sh [hunk 15]
```diff
@@ -24,8 +24,36 @@ rlJournalStart
         assert_up_and_running
     rlPhaseEnd
 
+    rlPhaseStartTest "Should not display SSL warning for HTTPS connection"
+        rlRun -t -c "docker exec -i kiwi_web /Kiwi/manage.py createsuperuser \
+            --username testadmin --email testadmin@domain.com --noinput"
+        rlRun -t -c "cat tests/set_testadmin_pass.py | docker exec -i kiwi_web /Kiwi/manage.py shell"
+        IP_ADDRESS=`docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' kiwi_web`
+        rm -f /tmp/testcookies.txt
+        rm -f /tmp/testdata.sslon
```
Comment (atodorov) [inline tests/test_http.sh:33]: Do you really need to remove those files or they get overwritten ?

File: tests/set_testadmin_pass.py (hunk 12 above)
Comment (atodorov) [inline tests/set_testadmin_pass.py:8]: This file makes black unhappy. Fix it.

File: tests/test_http.sh [hunk 16]
```diff
@@ -24,8 +24,36 @@ rlJournalStart
...
+        rlRun -t -c "curl -k -L -o- -c /tmp/testcookies.txt https://$IP_ADDRESS:8443/"
+        CSRF_TOKEN=`grep csrftoken /tmp/testcookies.txt | cut -f 7`
+        rlRun -t -c "curl -e https://$IP_ADDRESS:8443/accounts/login/ -d username=testadmin \
```
Comment (atodorov) [inline tests/test_http.sh:36]: see ";auto" for referrer. It could work and save you some typing.

File: tests/test_http.sh (hunk 14 above)
Comment (atodorov) [reply tests/test_http.sh:41]: ping ^^^

**Review by atodorov — CHANGES_REQUESTED** [review]
Minor chabnges requested.

File: tests/test_http.sh [hunk 17]
```diff
@@ -4,12 +4,21 @@
...
     sleep 10
-    IP_ADDRESS=`docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' kiwi_web`
     # both HTTP and HTTPS display the login page
     rlRun -t -c "curl    -L -o- http://$IP_ADDRESS:8080/  | grep 'Welcome to Kiwi TCMS'"
     rlRun -t -c "curl -k -L -o- https://$IP_ADDRESS:8443/ | grep 'Welcome to Kiwi TCMS'"
 }
 
+get_dashboard() {
+    rm -f /tmp/testcookies.txt
+    rlRun -t -c "curl -k -L -o- -c /tmp/testcookies.txt $URL/"
```
Comment (atodorov) [inline tests/test_http.sh:14]: Pass the URL variable as parameter to the function.

File: tests/test_http.sh [hunk 18]
```diff
@@ -4,12 +4,21 @@
 
 assert_up_and_running() {
     sleep 10
-    IP_ADDRESS=`docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' kiwi_web`
```
Comment (atodorov) [inline tests/test_http.sh:7]: Pass the IP_ADDRESS value as parameter to the function.

**Review by atodorov — CHANGES_REQUESTED** [review]
Minor changes requested, tests started failing again.

File: tests/test_http.sh (hunk 16 above)
Comment (schwarzkrieger) [reply tests/test_http.sh]: This seems to break HTTPS test. HTTP works fine.

File: tests/test_http.sh (hunk 15 above)
Comment (schwarzkrieger) [reply tests/test_http.sh]: It's not clearly stated in the curl documentation. Currently the output file is always overwritten, the cookie file is created / overwritten when -c option is used.

**Pull request merged**